### PR TITLE
Retrieve AccountNumber from metadata instead of GetUser

### PR DESCRIPTION
**Pull request by EduardSergeev**
Should work in the current IAM Role-based environment

Fixes https://github.com/ambiata/mismi/issues/278

### CODE REVIEW

**Comment by olorin** [comment]
Nice, I like this more than the last method. Not totally sure about having the account number in the tests though - @markhibberd can comment on that?

**Comment by EduardSergeev** [comment]
> Not totally sure about having the account number in the tests though

Yes, wasn't sure about it either. Maybe there is a better way to check that the number we get makes sense?

File: mismi-ec2/src/Mismi/EC2/Data.hs [hunk 1]
```diff
@@ -65,3 +69,23 @@ newtype SecurityGroupId =
...
+  } deriving (Eq, Show)
+
+newtype AccountNumber = AccountNumber {
+    unAccountNumber :: Text
+  } deriving (Eq, Show)
+
+extractAccountNumber :: InstanceProfileARN -> Maybe AccountNumber
+extractAccountNumber =
+  -- arn:partition:service:region:account-id:resourcetype/resource
+  fmap AccountNumber . listToMaybe . drop 4 . splitOn ":" . unInstanceProfileARN
```
Comment (olorin) [inline mismi-ec2/src/Mismi/EC2/Data.hs]: Should this validate that the "number" received is numeric?

**Comment by olorin** [comment]
> Yes, wasn't sure about it either. Maybe there is a better way to check that the number we get makes sense?

I can't think of a good one. Maybe just verifying that the number itself looks sane? (i.e., is a number)

File: mismi-ec2/src/Mismi/EC2/Data.hs (hunk 1 above)
Comment (EduardSergeev) [inline mismi-ec2/src/Mismi/EC2/Data.hs]: If it's guaranteed that AWS Account ID will always consists of digits only it isn't problem to add this additional check.

Comment (olorin) [inline mismi-ec2/src/Mismi/EC2/Data.hs]: Yep, it's a number. Looks like it's even [always twelve digits](https://docs.aws.amazon.com/general/latest/gr/acct-identifiers.html).

**Comment by EduardSergeev** [comment]
Bump @charleso @nhibberd  Instead of https://github.com/ambiata/mismi/pull/254 which requires `GetUser` permissions and doesn't work in the current environment. Good to be merged?

**Comment by nhibberd** [comment]
@markhibberd wanted to look at this

**Comment by markhibberd** [comment]
I missed this when it was added last time. I don't think we should be hitting an API to retrieve static information, multiple reasons, main ones being:
 - We already hit the APIs too much, to the point of self-denial-of-servicing ourselves, and with the request budget, which is as low as 100 requests a second per account, we just can't rely on them being available.
 - I am not ok that a permission was opened up to do this without checking with me. These things need to be thought through more, because we really don't want to be accidentally leaking information.

This PR obviously addresses these two issues somewhat by switching to the meta-data tree. But even that hasn't been ultra-reliable in the past, so not too enthused about relying on it for something like this (given it may need to happen a lot for things like queue names). As for the change I don't mind whether it goes into mismi or not just as a useful function to have, but I don't want this used downstream until other options are discussed. I will put up an issue on payload to discuss what I think is a reasonable solution.

**Comment by EduardSergeev** [comment]
I somehow missed in the original discussion the option [to get Account Number from environment](https://github.com/ambiata/anatomy-ambiata/pull/61#issuecomment-160835878). Which definitely looks like a better option to me now. I am happy to change this PR to just remove all AccountRelated API calls and get it from environment variable once it is set up.

**Comment by markhibberd** [comment]
@EduardSergeev I am good with that :+1:

**Pull request merged**